### DISCUSSION TITLE
IO FontAtlas memory corrupted?

### DISCUSSION THREAD

**Discussion by SryEyes** [ocornut/imgui#6218] [New Users Build/Link/Run issues ONLY!]
Hey there im using ImGUI for over 1 year and didn't run into a single problem yet except for now.
Every method accessing the Fonts in the current IO runs into a problem.
-I upgraded to latest v1.89.3
-Im using ImGui_ImplWin32 and ImGui_ImplDX11
-The whole GUI gets loaded via dynamic link library based on the Platform the user is on

```cpp
void InitImGui()
{
	ImGui::CreateContext();
	ImGui_ImplWin32_Init(window);
	ImGui_ImplDX11_Init(pDevice, pContext);	
	ImGuiIO& io = ImGui::GetIO();
	io.Fonts->AddFontDefault();//triggers ASSERT in AddFontFromMemoryTTF   IM_ASSERT(!Locked && "Cannot modify a locked ImFontAtlas between NewFrame() and EndFrame/Render()!");
//every other Function from Fonts doesn't work either like Build, AddFontFromFileTTF and so on
}
```

```cpp
void MainLoop()
{
doonce(InitImGui());
ImGui_ImplDX11_NewFrame();
ImGui_ImplWin32_NewFrame();
ImGui::NewFrame();
ImGui::ShowFontSelector("lol"); //works
	ImGuiIO& io = ImGui::GetIO();
	ImGui::ShowFontAtlas(io.Fonts); //crashes since io.Fonts has no correct data
........
ImGui::EndFrame();
}
```

I have read the FAQ, the FONTS.md, the Changelog and searched the github issues. But i can't seem to find a fix for that problem.
And yes im able to read the assert but InitImGui is called before NewFrame thats why i dont get it.
I traced every call where the fontatlas is built / locked like GetTexDataAsRGBA32 and insured they dont get called before doing something with io.Fonts
I tried setting fontcfg.FontDataOwnedByAtlas = false; and passing an shared_fontatlas though CreateContext();.

So im 99% sure io.Fonts isn't having valid data and therefor im not able to access its values. But i dont get why im able to use the default Font when I dont call anything font related. So imGui internals itself has a valid Fontatlas and is able to use it but when I'm getting the Fontatlas through io.Fonts its rubbish? Though of some memory error like maybe IM_NEW and dll's don't like each other? But then why is the default font built correctly?
I know its probs an error with my specific setup but does anyone have an idea what im able to try next? Is the default font when nothing is changed in io.Fonts built with the default fontatlas?

Screenshots are attached (dev envoirement is german but basically it tells that memory cant be read)

Thanks for your time
-Edge
![fa](https://user-images.githubusercontent.com/102036089/222752233-50ca106c-230c-4e00-a544-3b55a4cd530b.jpg)

**Comment by PathogenDavid** [comment]
> Edit:Sorry i tried to add this as an code block but then its all one line. I cant seem to fix that format so im posting it like that.

You need to use triple-backticks for multiple lines. IE:

~~~
```cpp
// Say hello
printf("Hello, world!");
```
~~~

becomes

```cpp
// Say hello
printf("Hello, world!");
```

-----------------

> ```cpp
> io.Fonts->AddFontDefault();//triggers ASSERT in AddFontFromMemoryTTF IM_ASSERT(!Locked && "Cannot modify a locked ImFontAtlas between NewFrame() and EndFrame/Render()!");
> ```

This assert should not be ignored. It sounds like your app isn't calling things in the correct order and you need to debug it to figure out what's wrong with it.

A good strategy would be to place a breakpoint in `ImGui::NewFrame`, `ImGui::EndFrame`, and `ImFontAtlas::AddFontFromMemoryTTF`. Run your app and use the stack traces to figure out why your calls are happening in the incorrect order.

**Reply by SryEyes** [reply]
Thanks for the response!
I traced it but the order seems fine which i dont understand.

Here is the callstack when i don't access io.fonts at all and only the default font gets loaded - so everything works

```
ImGui_ImplWin32_Init
ImGui_ImplDX11_Init
ImGui_ImplDX11_NewFrame
	which Builds the fontatlas and calls AddFontDefault and then AddFontFromMemoryTTF
ImGui_ImplWin32_NewFrame
ImGui::NewFrame
ImGui::Render
	which calls ImGui::EndFrame
ImGui_ImplDX11_RenderDrawData
```

And here is the the one where i only call io.Fonts->AddFontDefault(); after ImGui_ImplWin32_Init and ImGui_ImplDX11_Init

```
ImGui_ImplWin32_Init
ImGui_ImplDX11_Init
AddFontDefault which calles AddFontFromMemoryTTF
	which throws the IM_ASSERT(!Locked && "Cannot modify a locked ImFontAtlas between NewFrame() and EndFrame/Render()!");
```
Is there anything else i could try or possibly been missing? Context is created, Init functions are called and no Frame function is called yet but only adding the default font in seemingly the right place throws the asset.

I also tried to set dx11 in debug mode and ran my app without touching io.Fonts myself so the callstack was like in the first snippet above.
And then the app crashed in ImGui_ImplDX11_CreateFontsTexture at bd->pd3dDevice->CreateShaderResourceView which might be the error why im not able to call anything font related myself. But then why does it work when dx11 is not in debug mode? Maybe imgui doesn't like the debug version of dx11?

thanks
-SryEyes

**Reply by PathogenDavid** [reply]
What does your call to `ImGui::CreateContext` look like?

> Maybe imgui doesn't like the debug version of dx11?

Dear ImGui absolutely supports the debug layer.

**Reply by SryEyes** [reply]
Its just calling CreateContext without params right at the start of the InitImgui function like
```cpp
ImGuiContext* context = ImGui::CreateContext();
	ImGui_ImplWin32_Init(window);
	ImGui_ImplDX11_Init(pDevice, pContext);	
```
I also tried Creating a context with an shared_atlas but it didn't work either.

> Dear ImGui absolutely supports the debug layer.
Ok thanks for the info. Need to investigate that then C:

**Reply by PathogenDavid** [reply]
> I also tried Creating a context with an shared_atlas but it didn't work either.

That was actually why I asked.

The *only* place in Dear ImGui that locks the font atlas is [`ImGui::NewFrame`](https://github.com/ocornut/imgui/blob/c9a53aa74d84cdb6471651f7b3da14d7248d58c7/imgui.cpp#L4372), and [`ImFontAtlas` is 0-initialized upon construction](https://github.com/ocornut/imgui/blob/c9a53aa74d84cdb6471651f7b3da14d7248d58c7/imgui_draw.cpp#L1970) (so if you were passing a `shared_atlas` that might be suspect.)

So `ImGui::NewFrame` is being called *somewhere* or something unrelated (most likely in your codebase) is corrupting `ImFontAtlas`.

There is a non-zero chance that you're unintentionally accessing two separate contexts of Dear ImGui due to using it via a DLL. That can be pretty trick to investigate though. If your build setup is flexible enough it might be worth statically linking Dear ImGui and any code that accesses it in order to rule that out.

You should also be on the lookout for any multithreading issues. Dear ImGui is not threadsafe.

> And then the app crashed in ImGui_ImplDX11_CreateFontsTexture at bd->pd3dDevice->CreateShaderResourceView

Is anything written to the debug output window in Visual Studio? (Go to `View` > `Output` if the pane is missing and ensure `Show output from` is set to `Debug`. That's where the debug layer prints its messages.)

It would also be worth checking whether the values for `pixels`, `width`, and `height` look normal within `ImGui_ImplDX11_CreateFontsTexture`.

**Reply by SryEyes** [reply]
Bet this is a mulithreading issue. Ill try disabling multithreading and checking again.

Enabled the debug layer again and got this error in ImGui_ImplDX11_CreateFontsTexture:
ERROR! D3D call failed! hr = E_INVALIDARG ( D3D11Device()->CreateTexture2D( &desc, bImmutable ? pSubResource : nullptr, ( ID3D11Texture2D ** ) &pDx11Texture ) )

The app my addon is for itself is currently under heavy developement so maybe thats an issue which is on their side.

**Reply by PathogenDavid** [reply]
> ERROR! D3D call failed! hr = E_INVALIDARG ( D3D11Device()->CreateTexture2D( &desc, bImmutable ? pSubResource : nullptr, ( ID3D11Texture2D ** ) &pDx11Texture ) )

Are you sure that's coming from `ImGui_ImplDX11_CreateFontsTexture`? The message doesn't look like it's coming from the debug layer and the code given doesn't match the backend. That looks like it's maybe an assert from the app hosting your addon.

The debug layer messages should also tell you which parameter is wrong, for example if I force `desc.Width` to 0:

```
D3D11 ERROR: ID3D11Device::CreateTexture2D: The Dimensions are invalid. For feature level D3D_FEATURE_LEVEL_11_0, the Width (value = 0) must be between 1 and 16384, inclusively. The Height (value = 64) must be between 1 and 16384, inclusively. And, the ArraySize (value = 1) must be between 1 and 2048, inclusively. [ STATE_CREATION ERROR #101: CREATETEXTURE2D_INVALIDDIMENSIONS]
D3D11 ERROR: ID3D11Device::CreateTexture2D: Returning E_INVALIDARG, meaning invalid parameters were passed. [ STATE_CREATION ERROR #104: CREATETEXTURE2D_INVALIDARG_RETURN]
```

**Comment by SryEyes** [comment]
Just was able to get a custom font working.
Thanks for the help!
Figured out by looking at the callstack: when i dont access io.fonts ImFontAtlas::Build calls AddFontDefault since ConfigData.Size == 0. So i tried calling AddFontFromFileTTF right in ImFontAtlas::Build and didn't get an assert thrown. After that i jumped a few calls higher in the stack and called a bunch of io.Fonts stuff in ImGui_ImplDX11_CreateFontsTexture still without crashing or getting an assert. Now im just initializing my font stuff like this:

```cpp
static void ImGui_ImplDX11_CreateFontsTexture()
{
    // Build texture atlas
    ImGuiIO& io = ImGui::GetIO();

    io.Fonts->AddFontDefault();
    auto myfont= io.Fonts->AddFontFromFileTTF("....\\arial.ttf", 14.0f);
    ImGui_ImplDX11_Data* bd = ImGui_ImplDX11_GetBackendData();
    .....
```

Not a great way of doing it especially when it comes to getting a new update of imgui but it works for now.
Im really curious why it doesn't work outside the imgui code even tho the order of calls seems right.
Is it safe to add fonts in there since its before GetTexDataAsRGBA32? Im currently testing and didn't experience any bug yet.
I'll make a new post if i am able to figure out why i'm not able to access fonts the normal way or why CreateShaderResourceView is crashing in the debuglayer.
Thank you
-SryEyes

**Reply by PathogenDavid** [reply]
I think you've basically just side-stepped the issue by ensuring the font initialization only happens in a place where it's safe. (You basically just moved the call inside `ImGui_ImplDX11_NewFrame` which should be happening before `ImGui::NewFrame`.)

If you want to check if it's a multi-threading issue it'd be worth putting a breakpoint at your broken font code location and another within `ImGui_ImplDX11_CreateFontsTexture` and checking the thread ID for each.

**Reply by SryEyes** [reply]
Checked and both threads are the same (not the mainthread - but the same ones).
Is there any other reason the fontatlas could get locked besides being in a frame?

**Reply by PathogenDavid** [reply]
Unfortunately not. There's only three places `Locked` is written and two of them are setting it to false:

In `ImGui::NewFrame`:

https://github.com/ocornut/imgui/blob/a322122f74a2aa625a128e47e1e29fb2e304bb8b/imgui.cpp#L4378-L4379

In `ImGui::EndFrame`:

https://github.com/ocornut/imgui/blob/a322122f74a2aa625a128e47e1e29fb2e304bb8b/imgui.cpp#L4909-L4910

And in `ImGui::Shutdown`:

https://github.com/ocornut/imgui/blob/a322122f74a2aa625a128e47e1e29fb2e304bb8b/imgui.cpp#L3521

---------

Another thing you might double check is whether there's unintentionally multiple contexts involved here by checking the value of `ImGui::GetCurrentContext()` at various points.
